### PR TITLE
[FLINK-17448][sql-parser][table-api-java][table-planner-blink][hive] Implement table DDLs for Hive dialect part2

### PR DESCRIPTION
**Pull request by lirui-apache**
## What is the purpose of the change

To implement ALTER TABLE [PARTITION] for Hive dialect.

## Brief change log

  - Implement the syntax
  - Update `SqlAlterTable` to return an optional partition spec if the operation should be applied to partitions
  - Add `AlterPartitionOperation` and support it in the blink planner.
  - Most ALTER semantics are encoded as table properties. Update HiveCatalog to support that.
  - Add test cases.

## Verifying this change

Existing and added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

**Comment by lirui-apache** [comment]
The azure build passed for my personal repo: https://dev.azure.com/lirui-apache/flink/_build/results?buildId=35&view=results

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 1]
```diff
@@ -138,6 +158,14 @@
 	private static final String FLINK_FUNCTION_PREFIX = "flink:";
 	private static final String FLINK_PYTHON_FUNCTION_PREFIX = FLINK_FUNCTION_PREFIX + "python:";
 
+	private static final Pattern DECIMAL_PATTERN = Pattern.compile("^decimal(\\((\\d+),\\s*(\\d+)\\))?$");
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:161]: Move them to a type parser class.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 2]
```diff
@@ -1450,4 +1489,253 @@ static boolean isGenericForGet(Map<String, String> properties) {
...
+
+	private static void addReplaceColumns(StorageDescriptor sd, List<FieldSchema> columns, boolean replace) {
+		if (replace) {
+			sd.setCols(columns);
+		} else {
+			sd.getCols().addAll(columns);
+		}
+	}
+
+	private static TypeInfo calTypeStrToHiveTypeInfo(String typeStr) {
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:1633]: Move them to a type parser class.

Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:1633]: Is there no util from hive or flink-sql-parser-hive?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java [hunk 3]
```diff
@@ -484,18 +512,24 @@ public void alterTable(ObjectPath tablePath, CatalogBaseTable newCatalogTable, b
...
-
-		// client.alter_table() requires a valid location
-		// thus, if new table doesn't have that, it reuses location of the old table
-		if (!newTable.getSd().isSetLocation()) {
-			newTable.getSd().setLocation(hiveTable.getSd().getLocation());
+		boolean isGeneric = isGenericForGet(hiveTable.getParameters());
+		if (isGeneric) {
+			hiveTable = alterTableViaCatalogBaseTable(tablePath, newCatalogTable, hiveTable);
+		} else {
+			AlterTableOp op = extractAlterTableOp(newCatalogTable.getOptions());
```
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:519]: Move them to an alter util?
Or maybe a `AlterTable`, and have two implementation: `ObjectAlterTable` and `PropertiesAlterTable`.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java (hunk 2 above)
Comment (JingsongLi) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:1633]: Add a `AlterTableColumnTypeOperation`?

**Comment by lirui-apache** [comment]
I have updated the PR to add operations for the ALTER COLUMNS statements, so that we don't have to store column types as table properties.

With this approach, we cannot support ALTER COLUMNS for a single partition, because `CatalogPartition` isn't associated with a schema. But users can still cascade the change to all partitions.

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java (hunk 3 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/HiveCatalog.java:496]: I can extract some methods to the util class, but it's difficulty to move all of them. The `alterTableViaProperties` needs to access the HMS client to do the cascade ALTER COLUMNS. So I don't think we should make it static.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+/**
+ * Utils methods for converting sql to operations.
+ */
+public class OperationConverterUtils {
+
+	private OperationConverterUtils() {
+	}
+
+	public static Operation convertAddReplaceColumns(ObjectIdentifier tableIdentifier,
+			SqlAddReplaceColumns addReplaceColumns, CatalogTable catalogTable, SqlValidator sqlValidator) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:61]: one parameter one line.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		// verify partitions columns appear last in the schema
+		TableSchema oldSchema = catalogTable.getSchema();
+		int numPartCol = catalogTable.getPartitionKeys().size();
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:64]: This is not flink partition style, this is hive partition style

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/SqlPartitionUtils.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	/** Get static partition key value pair as strings.
+	 *
+	 * <p>For character literals we return the unquoted and unescaped values.
+	 * For other types we use {@link SqlLiteral#toString()} to get
+	 * the string format of the value literal.
+	 *
+	 * @return the mapping of column names to values of partition specifications,
+	 * returns an empty map if there is no partition specifications.
+	 */
+	public static LinkedHashMap<String, String> getPartitionKVs(SqlNodeList partitionSpec) {
```
Comment (JingsongLi) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/SqlPartitionUtils.java:45]: `getPartitionSpec`?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 7]
```diff
@@ -231,82 +239,102 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
...
+			CatalogPartition catalogPartition = catalogManager.getPartition(tableIdentifier, partitionSpec)
+					.orElseThrow(() -> new ValidationException(String.format("Partition %s of table %s doesn't exist",
+							partitionSpec.getPartitionSpec(), tableIdentifier)));
+			Map<String, String> props = catalogPartition.getProperties();
+			alterTableProperties.getPropertyList().getList().forEach(p ->
+					props.put(((SqlTableOption) p).getKeyString(), ((SqlTableOption) p).getValueString()));
+			return new AlterPartitionPropertiesOperation(tableIdentifier, partitionSpec, catalogPartition);
+		}
+		// it's altering a table
+		if (baseTable instanceof CatalogTable) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:324]: else if

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 8]
```diff
@@ -231,82 +239,102 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
...
+			Map<String, String> newProperties = new HashMap<>(oldTable.getProperties());
+			newProperties.putAll(OperationConverterUtils.extractProperties(alterTableProperties.getPropertyList()));
+			CatalogTable newTable = new CatalogTableImpl(
+					oldTable.getSchema(),
+					oldTable.getPartitionKeys(),
+					newProperties,
+					oldTable.getComment());
+			return new AlterTablePropertiesOperation(tableIdentifier, newTable);
+		}
+		throw new ValidationException("Unsupported CatalogBaseTable type: " + baseTable.getClass().getName());
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:335]: else

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 9]
```diff
@@ -231,82 +239,102 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
...
+			alterTableProperties.getPropertyList().getList().forEach(p ->
+					props.put(((SqlTableOption) p).getKeyString(), ((SqlTableOption) p).getValueString()));
+			return new AlterPartitionPropertiesOperation(tableIdentifier, partitionSpec, catalogPartition);
+		}
+		// it's altering a table
+		if (baseTable instanceof CatalogTable) {
+			CatalogTable oldTable = (CatalogTable) baseTable;
+			Map<String, String> newProperties = new HashMap<>(oldTable.getProperties());
+			newProperties.putAll(OperationConverterUtils.extractProperties(alterTableProperties.getPropertyList()));
+			CatalogTable newTable = new CatalogTableImpl(
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:328]: `oldTable.copy`

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java [hunk 10]
```diff
@@ -231,82 +239,102 @@ private Operation convertDropTable(SqlDropTable sqlDropTable) {
...
+	private Operation convertAlterTableProperties(ObjectIdentifier tableIdentifier, CatalogBaseTable baseTable,
+			SqlAlterTableProperties alterTableProperties) {
+		LinkedHashMap<String, String> partitionKVs = alterTableProperties.getPartitionKVs();
+		// it's altering partitions
+		if (partitionKVs != null) {
+			CatalogPartitionSpec partitionSpec = new CatalogPartitionSpec(partitionKVs);
+			CatalogPartition catalogPartition = catalogManager.getPartition(tableIdentifier, partitionSpec)
+					.orElseThrow(() -> new ValidationException(String.format("Partition %s of table %s doesn't exist",
+							partitionSpec.getPartitionSpec(), tableIdentifier)));
+			Map<String, String> props = catalogPartition.getProperties();
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/operations/SqlToOperationConverter.java:318]: Should create a new `CatalogPartition` with a new Map.

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/SqlPartitionUtils.java (hunk 6 above)
Comment (lirui-apache) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/SqlPartitionUtils.java:45]: In our name convention, a partition spec is usually a `SqlNodeList`. And this is following the example of `RichSqlInsert`, whose `getStaticPartitionKVs` returns a map.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java (hunk 5 above)
Comment (lirui-apache) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:70]: I have added some explanations in the comments.

**Pull request merged**